community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexMap.java

```diff
@@ -95,32 +130,204 @@ public void foreachIndexProxy( BiConsumer<Long, IndexProxy> consumer )
+    }
+
     @Override
     public IndexMap clone()
     {
         return new IndexMap( cloneMap( indexesById ), cloneMap( indexesByDescriptor ),
                 cloneMap( indexIdsByDescriptor ) );
     }
 
+    public Iterator<LabelSchemaDescriptor> descriptors()
+    {
+        return indexesByDescriptor.keySet().iterator();
+    }
+
+    public Iterator<Long> indexIds()
+    {
+        return indexesById.keySet().iterator();
+    }
+
+    public int size()
+    {
+        return indexesById.size();
+    }
+
+    // HELPERS
+
     private <K, V> Map<K, V> cloneMap( Map<K, V> map )
     {
         Map<K, V> shallowCopy = new HashMap<>( map.size() );
         shallowCopy.putAll( map );
         return shallowCopy;
     }
 
-    public Iterator<LabelSchemaDescriptor> descriptors()
+    private void addDescriptorToLookups( LabelSchemaDescriptor schema )
     {
-        return indexesByDescriptor.keySet().iterator();
+        addToLookup( schema.getLabelId(), schema, descriptorsByLabel );
+
+        for ( int propertyId : schema.getPropertyIds() )
+        {
+            addToLookup( propertyId, schema, descriptorsByProperty );
+        }
     }
 
-    public Iterator<Long> indexIds()
+    private void addToLookup(
+            int key,
+            LabelSchemaDescriptor schema,
+            PrimitiveIntObjectMap<Set<LabelSchemaDescriptor>> lookup )
     {
-        return indexesById.keySet().iterator();
+        Set<LabelSchemaDescriptor> descriptors = lookup.get( key );
+        if ( descriptors == null )
+        {
+            descriptors = new HashSet<>();
+            lookup.put( key, descriptors );
+        }
+        descriptors.add( schema );
     }
 
-    public int size()
+    private void removeFromLookup(
+            int key,
+            LabelSchemaDescriptor schema,
+            PrimitiveIntObjectMap<Set<LabelSchemaDescriptor>> lookup )
     {
-        return indexesById.size();
+        Set<LabelSchemaDescriptor> descriptors = lookup.get( key );
+        descriptors.remove( schema );
+        if ( descriptors.isEmpty() )
+        {
+            lookup.remove( key );
+        }
+    }
+
+    private static Map<LabelSchemaDescriptor, IndexProxy> indexesByDescriptor( Map<Long,IndexProxy> indexesById )
+    {
+        Map<LabelSchemaDescriptor, IndexProxy> map = new HashMap<>();
+        for ( IndexProxy proxy : indexesById.values() )
+        {
+            map.put( proxy.schema(), proxy );
+        }
+        return map;
+    }
+
+    private static Map<LabelSchemaDescriptor, Long> indexIdsByDescriptor( Map<Long,IndexProxy> indexesById )
+    {
+        Map<LabelSchemaDescriptor, Long> map = new HashMap<>();
+        for ( Map.Entry<Long,IndexProxy> entry : indexesById.entrySet() )
+        {
+            map.put( entry.getValue().schema(), entry.getKey() );
+        }
+        return map;
+    }
+
+    /**
+     * Get descriptors affected by changed properties. Implementation is smart about whether doing
```

Comment:
would be nice to have a description what `smart` actually mean and how those lookups will be done